flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/functions/hive/HiveScalarFunction.java

```diff
@@ -0,0 +1,107 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.functions.hive;
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.api.common.typeinfo.TypeInformation;
+import org.apache.flink.table.functions.FunctionContext;
+import org.apache.flink.table.functions.ScalarFunction;
+import org.apache.flink.table.types.CollectionDataType;
+import org.apache.flink.table.types.DataType;
+import org.apache.flink.table.types.utils.LegacyTypeInfoDataTypeConverter;
+
+import org.apache.hadoop.hive.ql.exec.UDF;
+import org.apache.hadoop.hive.ql.udf.generic.GenericUDF;
+import org.apache.hadoop.hive.serde2.objectinspector.ObjectInspector;
+
+
+/**
+ * Abstract class to provide more information for Hive {@link UDF} and {@link GenericUDF} functions.
+ */
+@Internal
+public abstract class HiveScalarFunction<UDFType> extends ScalarFunction implements HiveFunction {
+
+	protected final HiveFunctionWrapper<UDFType> hiveFunctionWrapper;
+
+	protected Object[] constantArguments;
+	protected DataType[] argTypes;
+
+	protected transient UDFType function;
```

Comment:
@xuefuz sorry that I accidentally deleted your comment here. So your question is what's the difference between this function instance and the one in `HiveFunctionWrapper`.

The one in `HiveFunctionWrapper` acts as a cache for GenericUDF, GenericUDTA, and UDAFResolver, but not `UDF` (I'm still not sure why not `UDF`, that logic is borrowed from Spark). So this function instance in `HiveScalarFunction` is actually the one gonna be invoked to process data.

Reply:
okay.

Reply:
I think the reason is in Hive `UDF` comments:
```
One instance of this class will be instantiated per JVM and it will not be called concurrently.
```